tests/robustness/traffic/kubernetes.go

```diff
@@ -102,54 +94,92 @@ func (t kubernetesTraffic) Run(ctx context.Context, c *RecordingClient, limiter
 			}
 			// Avoid multiple failed writes in a row
 			if lastWriteFailed {
-				listCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
-				resp, err := kc.List(listCtx, keyPrefix)
-				cancel()
+				_, err := t.Read(ctx, kc, s, limiter, keyPrefix)
 				if err != nil {
 					continue
 				}
-				s.Reset(resp)
-				limiter.Wait(ctx)
 			}
-			writeCtx, cancel := context.WithTimeout(ctx, RequestTimeout)
-			err := t.Write(writeCtx, kc, ids, s)
-			cancel()
+			err := t.Write(ctx, kc, ids, s, limiter)
 			lastWriteFailed = err != nil
 			if err != nil {
 				continue
 			}
-			limiter.Wait(ctx)
 		}
 	})
 	g.Wait()
 }
 
-func (t kubernetesTraffic) Write(ctx context.Context, kc *kubernetesClient, ids identity.Provider, s *storage) (err error) {
+func (t kubernetesTraffic) Read(ctx context.Context, kc *kubernetesClient, s *storage, limiter *rate.Limiter, keyPrefix string) (rev int64, err error) {
+	limit := int64(t.averageKeyCount)
```

Comment:
umm, can we start with a smaller limit to guarantee the following pagination?

Reply:
Based on experiments this is enough to guarantee pagination, any less there is too much requests.

Reply:
> We get one range request per two lists.

I am not sure I understand this statement. I would expect multiple range requests for each list. This should fall into the second goroutine. 

First `Read` is used in the first goroutine which is ListAndWatch in kubernets. It's a full scan of the resource key prefix range. 

The second `Read` in another goroutine is k8s paginated list requests, each page count is 500 in kubernetes by default. Usually it results in client side pagination.